### ISSUE TITLE
Question about condition Check for Relative Drawdown in MaxDrawDownRelativeHyperOptLoss

### ISSUE THREAD

**Issue by keithorange** [freqtrade/freqtrade#8694]
## Description
The `hyperopt_loss_function` within the `MaxDrawDownRelativeHyperOptLoss` class contains an error in the condition check for `relative_drawdown`. This issue affects the accuracy of the loss calculation, potentially leading to suboptimal optimization results.

## Error
In the current implementation, the condition `if max_drawdown == 0` is used to determine if `relative_drawdown` should be considered in the loss calculation. However, this condition does not account for cases where `relative_drawdown` is extremely small but not exactly zero due to floating-point precision limitations. As a result, the loss function may incorrectly exclude cases where `relative_drawdown` is close to zero but still relevant for the optimization process.

## Solution
To address this issue, it is recommended to modify the condition check to consider a tolerance level for `relative_drawdown`. By comparing the absolute value of `relative_drawdown` with a small tolerance, we can handle cases where `relative_drawdown` is very close to zero but not exactly zero due to floating-point precision limitations.

Here is the updated code with the proposed solution implemented:

```python
class MaxDrawDownRelativeHyperOptLoss(IHyperOptLoss):
    """
    Defines the loss function for hyperopt.

    This implementation optimizes for max drawdown and profit.
    Lower max drawdown and higher profit yield lower return values.
    """

    @staticmethod
    def hyperopt_loss_function(results: DataFrame, config: Config,
                               *args, **kwargs) -> float:
        """
        Objective function.

        Uses profit ratio weighted max_drawdown when drawdown is available.
        Otherwise, directly optimizes profit ratio.
        """
        total_profit = results['profit_abs'].sum()
        try:
            drawdown_df = calculate_underwater(
                results,
                value_col='profit_abs',
                starting_balance=config['dry_run_wallet']
            )
            max_drawdown = abs(min(drawdown_df['drawdown']))
            relative_drawdown = max(drawdown_df['drawdown_relative'])
            tolerance = 1e-2  # Adjust the tolerance as needed

            if max_drawdown == 0 or abs(relative_drawdown) < tolerance:
                return -total_profit

            return -total_profit / (max_drawdown * relative_drawdown)
        except (Exception, ValueError):
            return -total_profit
```

By incorporating the tolerance level and checking `abs(relative_drawdown) < tolerance`, the condition properly accounts for small values of `relative_drawdown`. This ensures accurate loss calculation and improves the reliability of the optimization process.

## Impact and Importance
Fixing this issue is crucial to ensure the accuracy and effectiveness of the hyperparameter optimization process. By correctly considering `relative_drawdown` in the loss calculation, the optimization algorithm can make more informed decisions and identify optimal hyperparameters that yield lower max drawdown and higher profit. This leads to improved performance and better decision-making in trading strategies.

Please consider the updated description and solution for the issue. If you have any further questions or need additional assistance, feel free to let me know.

**Comment by xmatthias** [comment]
If you think things should be improved and have a solution ready, please submit that as a pull request, not as an issue.
A guide for the github PR workflow can be found [here](https://guides.github.com/introduction/flow/).

A "solution" pasted in an issue is NOT something to work with easily. 

---

I'm not sure i agree with your analysis though. the `max_drawdown == 0` check is there to avoid 0 division errors - not necessarily to change the results per se.

max_drawdown = 0 is an edge-case which happens if there was no losing trade.
The moment you have at least **one** losing trade, this case won't be hit - but it'll follow the regular calculation (which in my opinion, is correct). 

Say you have a relative drawdown of 0.01% - then you'd calculate 1000$ (the profit) / 0.1 / 0.0001. - why would this have to be assumed "there was no drawdown" ? that's incorrect in my opinion.

Maybe you can highlight the problem with some calculation examples which would benefit from the modification you're proposing ? 

---

It's also unclear (and unexplained) why you're changing the calculation of `-total_profit / max_drawdown / relative_drawdown` to `-total_profit / (max_drawdown * relative_drawdown)`.
isn't that mathematically identical (except for some rounding errors maybe) - unless i'm missing something?

**Comment by keithorange** [comment]
@xmatthias 
ChatGPT Says:

The different drawdown metrics represent different ways of looking at the maximum loss, each of which can be useful depending on the context:

max_drawdown: This represents the maximum loss as a percentage of the account value. It's calculated as the maximum drop from a peak to a trough in the value of the account, divided by the account value at the peak. This is a commonly used measure of risk in finance.

max_drawdown_account: This represents the maximum loss in terms of the account size. It's the maximum value of the drawdown column divided by the account size, which is equivalent to max_drawdown.

max_relative_drawdown: This also represents the maximum loss in terms of the account size, so it should be the same as max_drawdown_account if the calculations are done correctly. However, it seems there might be a discrepancy in the calculations, as the numbers are not exactly the same. This could potentially be due to rounding errors or a difference in how the drawdown is calculated.

max_drawdown_abs: This represents the maximum loss in absolute terms, i.e., the total amount of money lost at the worst point. It's calculated as the maximum drop from a peak to a trough in the value of the account.

drawdown: This represents the maximum drawdown as a percentage of the account value at the worst point. It should be equivalent to max_drawdown if calculated correctly.

drawdown_relative: This is a metric you've created, which represents the total loss divided by both the maximum drawdown and the relative drawdown. It's a measure of the risk-adjusted return.

The fact that max_drawdown, max_drawdown_account, and max_relative_drawdown are not the same suggests there might be a discrepancy in the calculations. You might want to double-check the calculations for these metrics to make sure they're being done correctly.```

The fact that max_drawdown and drawdown are not the same also suggests a potential discrepancy in the calculations. Again, you might want to double-check the calculations to make sure they're being done correctly.


``` # OnlyProfitHyperOptLoss (OP): Considers only the amount of profit
            profit = CustomWeightSumHyperOptLoss.onlyprofit(results, trade_count, *args, **kwargs)

            # SharpeHyperOptLoss (SR): Optimizes the Sharpe Ratio
            sharpe_ratio = CustomWeightSumHyperOptLoss.sharpe(
                results, trade_count, config=config, *args, **kwargs)

            # SortinoHyperOptLoss (SOR): Optimizes the Sortino Ratio
            sortino_ratio = CustomWeightSumHyperOptLoss.sortino(
                results, trade_count, config=config, *args, **kwargs)

            # CalmarHyperOptLoss (CAL): Optimizes the Calmar Ratio
            calmar_ratio = CustomWeightSumHyperOptLoss.calmar(
                results, trade_count, config=config, *args, **kwargs)

            drawdown_relative = CustomWeightSumHyperOptLoss.drawdown_relative(
                results, config, *args, **kwargs)

            drawdown = CustomWeightSumHyperOptLoss.drawdown(
                results,  *args, **kwargs)

            # Include Backtest Stats
            # print(kwargs.keys())
            # print(kwargs['backtest_stats'].keys())

            stats = kwargs['backtest_stats']
            wins, draws, losses = stats["wins"], stats["draws"], stats["losses"]
            profit_mean, profit_median, profit_total = stats["profit_mean"], stats["profit_median"], stats['profit_total']
            # The Compound Annual Growth Rate (CAGR) reflects the annualized rate of return for your trading strategy. A higher CAGR indicates better performance over time.
            cagr = stats["cagr"]
            expectancy = stats["expectancy"]

            winner_holding_avg, winner_holding_avg_s = stats['winner_holding_avg'], stats[
                'winner_holding_avg_s']
            loser_holding_avg, loser_holding_avg_s = stats['loser_holding_avg'], stats[
                'loser_holding_avg_s']

            max_drawdown = stats['max_drawdown']
            max_drawdown_account = stats['max_drawdown_account']
            max_relative_drawdown = stats['max_relative_drawdown']
            max_drawdown_abs = stats['max_drawdown_abs']

            print("profit:", -profit)
            print("sharpe_ratio:", -sharpe_ratio)
            print("sortino_ratio:", -sortino_ratio)
            print("calmar_ratio:", -calmar_ratio)
            print("drawdown_relative:", -drawdown_relative)
            print("drawdown:", -drawdown)

            print("wins:", wins)
            print("draws:", draws)
            print("losses:", losses)
            print("profit_mean:", profit_mean)
            print("profit_median:", profit_median)
            print("profit_total:", profit_total)
            print("cagr:", cagr)
            print("expectancy:", expectancy)
            print("winner_holding_avg:", winner_holding_avg)
            print("winner_holding_avg_s:", winner_holding_avg_s)
            print("loser_holding_avg:", loser_holding_avg)
            print("loser_holding_avg_s:", loser_holding_avg_s)
            print("max_drawdown:", max_drawdown)
            print("max_drawdown_account:", max_drawdown_account)
            print("max_relative_drawdown:", max_relative_drawdown)
            print("max_drawdown_abs:", max_drawdown_abs)
```
```
profit: -9.643622350000001
sharpe_ratio: -8.046350543779873
sortino_ratio: -100
calmar_ratio: -496.22437117346374
drawdown_relative: -27981.880562447768
drawdown: -0.5194667796048076
wins: 1
draws: 1
losses: 1
profit_mean: -0.0032442669950736858
profit_median: 0.0
profit_total: -9.643622350000001e-06
cagr: -0.0017584217268583613
expectancy: -0.5064889265358685
winner_holding_avg: 0:33:00
winner_holding_avg_s: 1980.0
loser_holding_avg: 4:31:00
loser_holding_avg_s: 16260.0
max_drawdown: 0.018735627100520853
max_drawdown_account: 1.856429843081677e-05
max_relative_drawdown: 1.856429843081677e-05
max_drawdown_abs: 18.56446404 why are tgese dufferent max_drawdown: 0.018735627100520853
max_drawdown_account: 1.856429843081677e-05
max_relative_drawdown: 1.856429843081677e-05
max_drawdown_abs: 18.56446404
drawdown: -0.5194667796048076 drawdown_relative: max 18.56446404 
```

**Comment by keithorange** [comment]
The current drawdown_relative function you've written calculates a measure of the total loss divided by both the maximum drawdown and the relative drawdown. This is not a conventional financial metric and can lead to misleading results, especially if either the maximum drawdown or the relative drawdown is very small.

A more standard approach is to calculate the maximum drawdown relative to the peak value of the account, which measures the largest percentage drop in value from a peak. Here is a function that calculates this:
```
@staticmethod
def max_drawdown_relative(results: DataFrame, *args, **kwargs) -> float:
    """
    Calculate the maximum drawdown in the relative terms (percentage of the peak value)
    """
    # Cumulative profit
    cumulative_profit = results['profit_abs'].cumsum()

    # Running maximum
    running_max = np.maximum.accumulate(cumulative_profit)

    # Relative drawdown
    relative_drawdown = (running_max - cumulative_profit) / running_max

    # Maximum relative drawdown
    max_relative_drawdown = relative_drawdown.max()

    return max_relative_drawdown
```

This function should give you a value between 0 and 1, representing the largest percentage loss from a peak in the cumulative profit.

In general, when choosing a drawdown measure, it's important to consider the context. If you're trying to measure the risk of a trading strategy, you might want to consider both the maximum drawdown (which measures the worst-case loss) and the average drawdown (which measures the typical loss). If you're interested in risk-adjusted returns, you might want to consider a metric like the Calmar ratio, which is the annualized return divided by the maximum drawdown.

In any case, it's important to understand what each metric is measuring, and to make sure the calculations are done correctly.

**Comment by xmatthias** [comment]
@keithorange I assume both the above comments were completely written by chatgpt?

That makes every single word in there not trustworthy ans a waste of my time reading it. 
Parts of what the above comments say does not make any sense - so you probably didn't even read it, let alone think if what it says makes actually sense.
without the prompt chatgpt was given, it's also unclear what it's even dreaming about.

We have [documentation](https://www.freqtrade.io/en/stable/backtesting/#summary-metrics) which explains how each drawdown metric is to be interpreted. I'd suggest you to read that instead of asking chatgpt to get some bullshit answers.

At any rate - I consider chatGPT issue comments  as **spam** (they are wrong at any rate, and i'll not bother trying to educate a chatbot generated answer about that) - and will therefore mark both the above answers as such (i'm sorry if one of the 2 wasn't from chatGPT - but both sound like they were).

I'm however more than happy to get proper answers from you personally.
I'll however also have no problems to close an issue (irrelevant if the topic is correct or incorrect) if it's getting only spammy, autogenerated comments.

---

The great and funny part is that the function chatgpt wrote is actually exactly the calculation we're using (the only difference is that freqtrade is using pandas functions instead of numpy - which of course, neither you nor chatGPT actually checked.

**Comment by xmatthias** [comment]
As there was no follwup and chatgpt agreed with the current implementation, i assume this is no longer relevant, and your questions have been answered.

Feel free to comment below if that's not the case.
